### ISSUE TITLE
Misleading "unused" warning message

### ISSUE THREAD

**Issue by GuillaumeGomez** [rust-lang/rust#27559]
We got the following warning message:

```Shell
warning: variant is never used: `Foo`, #[warn(dead_code)] on by default
```

For this code:

```Rust
enum Something {
    Foo,
    NotFoo
}

fn main() {
    match Something::NotFoo {
        Something::Foo => {}
        _ => {}
    }
}
```

This warning could mislead users because for them, the Foo variant is used. Adding a specific warning message could be nice.

@eddyb proposed this message instead in this case:

"this variant is never instantiated"

cc @eddyb

**Comment by sanxiyn** [comment]
Closing as a duplicate of #19140.
